crates/pathfinder/src/storage/contract.rs

```diff
@@ -1,60 +1,83 @@
 use crate::{
     core::{ByteCodeWord, ContractAddress, ContractCode, ContractHash},
-    storage::{DB_VERSION_CURRENT, DB_VERSION_EMPTY},
+    storage::{DB_VERSION_1, DB_VERSION_CURRENT, DB_VERSION_EMPTY},
 };
 
 use anyhow::Context;
 use pedersen::StarkHash;
 use rusqlite::{named_params, OptionalExtension, Transaction};
 
-/// Migrates the [ContractsTable] to the [current version](DB_VERSION_CURRENT).
+/// Migrates the [ContractCodeTable] to the [current version](DB_VERSION_CURRENT).
 pub fn migrate(transaction: &Transaction, from_version: u32) -> anyhow::Result<()> {
+    ContractCodeTable::migrate(transaction, from_version)?;
     ContractsTable::migrate(transaction, from_version)
 }
 
 /// Stores StarkNet contract information, specifically a contract's
 ///
-/// - [address](ContractAddress)
 /// - [hash](ContractHash)
 /// - byte code
 /// - ABI
 /// - definition
-pub struct ContractsTable {}
+pub struct ContractCodeTable {}
 
-impl ContractsTable {
-    /// Migrates the [ContractsTable] from the given version to [DB_VERSION_CURRENT].
+impl ContractCodeTable {
+    /// Migrates the [ContractCodeTable] from the given version to [DB_VERSION_CURRENT].
     fn migrate(transaction: &Transaction, from_version: u32) -> anyhow::Result<()> {
+        const CREATE_CONTRACT_CODE_TABLE: &str = r"
+        CREATE TABLE contract_code (
+            hash       BLOB PRIMARY KEY,
+            bytecode   BLOB,
+            abi        BLOB,
+            definition BLOB
+        )";
+
         match from_version {
-            DB_VERSION_EMPTY => {} // Fresh database, continue to create table.
+            DB_VERSION_EMPTY => {
+                // Fresh database, just create the table.
+                transaction.execute(CREATE_CONTRACT_CODE_TABLE, [])?;
+            }
+            DB_VERSION_1 => {
+                // Rename the old contracts table
+                transaction.execute(r"ALTER TABLE contracts RENAME TO contracts_v1", [])?;
+                // Create the contract code table
+                transaction.execute(CREATE_CONTRACT_CODE_TABLE, [])?;
+                // Populate the contract code table with data
+                transaction.execute(
+                    r"INSERT OR IGNORE INTO contract_code (hash, bytecode, abi, definition)
```

Comment:
What does this `OR IGNORE` do here? Given the table was just created, surely there are no rows there, or does the select produce duplicate rows?

Reply:
Yes, it does produce duplicate rows.

Reply:
Added the duplicate rows case to the actual test as I discovered this when it blew in my face while migrating the big db dump.